core/trino-main/src/main/java/io/trino/operator/join/unspilled/JoinProbe.java

```diff
@@ -108,7 +103,64 @@ public Page getPage()
         return page;
     }
 
-    private boolean currentRowContainsNull()
+    private long[] fillCache()
+    {
+        long[] joinPositionCache = new long[positionCount];
+        Arrays.fill(joinPositionCache, -1);
+        if (probeMayHaveNull(probePage)) {
+            int nonNullCount = 0;
+            boolean[] isNull = new boolean[positionCount];
+            for (int i = 0; i < positionCount; i++) {
+                isNull[i] = rowContainsNull(i);
+                nonNullCount += isNull[i] ? 0 : 1;
+            }
+            if (nonNullCount < positionCount) {
+                // We only store positions that are not null
+                int[] positions = new int[nonNullCount];
+                nonNullCount = 0;
+                for (int i = 0; i < positionCount; i++) {
+                    if (!isNull[i]) {
+                        positions[nonNullCount++] = i;
+                    }
+                }
+                long[] packedPositionCache;
+                if (probeHashBlock != null) {
+                    long[] hashes = new long[nonNullCount];
+                    for (int i = 0; i < nonNullCount; i++) {
+                        hashes[i] = BIGINT.getLong(probeHashBlock, positions[i]);
+                    }
+                    packedPositionCache = lookupSource.getJoinPosition(positions, probePage, page, hashes);
+                }
+                else {
+                    packedPositionCache = lookupSource.getJoinPosition(positions, probePage, page);
+                }
+                // Unpack
+                nonNullCount = 0;
+                for (int i = 0; i < positionCount; i++) {
+                    if (!isNull[i]) {
+                        joinPositionCache[i] = packedPositionCache[nonNullCount++];
+                    }
+                }
+                return joinPositionCache;
+            } // else fall back to non-null path
+        }
+        int[] positions = new int[positionCount];
```

Comment:
Do you know the distribution of the expected `positionCount` values? If the usual case is larger than 1-2K or so, it may be beneficial to have a smaller batch size here to avoid trashing L1 cache with every loop over this and related arrays.

Reply:
A typical page would have somewhere between 1k-8k positions. However, all accesses to data here are sequential, so the data need not to be kept in the L1 cache.
We could obviously try it out but:
The underlying lookup source lookup source is usually a partitioned lookup source which means that the, let's say, 2k position page will be split into, let's say, 32 partitions. This way we end up with 64 positions per partition and the lower level lookup sources end up with batches of that size. With smaller pre-partition batches, the number of position per partition would be so low that the batching advantage may be diminished.
I hope I made myself clear, if not, please reach out offline